text/0078-improve-tso-scalability.md

```diff
@@ -0,0 +1,93 @@
+moment, PD client may receive 10 TSO requests at once, it will send just one TSO gRPC
+request to the PD leader with an argument `count` inside the gRPC request body.
+
+![TSO Client Batch - Old](https://i.imgur.com/vUgVSUI.png)
+
+For the PD server, it just takes the request and returns an incremental and unique TSO
+request with 4 counts and return it to the client. After the PD client receives the
+response, it will split the single TSO request into 10 different TSOs and return it to
+the upper requester.
+
+During the whole process flow, every PD client will maintain a gRPC stream with the PD
+leader to send and receive the TSO request efficiently.
+
+### Enhancement #1: Improve the batch effect
+
+An intuitive improvement is to improve the batch effect of TSO, i.e. to reduce the number
+of TSO gRPC requests by increasing the size of each batch with the same number of TSO requests.
+With fewer TSO gRPC requests, the CPU pressure on the PD leader can be reduced directly.
+
+We can introduce different batch strategies such as waiting for a while to fetch more TSO requests
+in a same interval or even more complicated one like a dynamic smallest batch size.
+
+![TSO Client Batch - New](https://i.imgur.com/HI6Tu2h.png)
+
+As the figure above shows, we can introduce a session variable like `@@tidb_tso_client_batch_max_wait_time`
+to control the max wait time the PD client is willing to wait for more TSO requests, which could make the
+batch size bigger without hurting the latency.
+
+Predicting strategy is also a useful way to improve the batch effect. For example, the PD Client
+could collect as much information as it needs such as latency and batch size in the last few minutes,
+then base on these information, the PD client could calculate a suitable expected batch size to predict
+the incoming TSO number, which make the batch waiting more effective.
+
+### Enhancement #2: Use proxy to reduce the stream number
+
+However, as mentioned before, according to our pprof result, the main reason of high PD leader
+CPU usage is the Go Runtime scheduling caused by the gRPC connection switch. Batch effect
+improvement does not alleviate this part at the root. To reduce the connection switch,
+fewer gRPC stream number is necessary. We can introduce the TSO Follower Proxy feature
+to achieve this.
+
+Every TSO request will be sent to different TSO servers (including both the PD leader and follower)
+randomly by the PD client, and multiple TSO requests sent to the same TSO follower will be
+batched again with the same logic as a PD client before being forwarded to the PD leader.
+With this implementation, the gRPC pressure is distributed to each PD server, and for the
+PD leader, if we have 50 TiDB instances and 5 PD instances, it only needs to maintain 4
+stream connections with each PD follower rather than 50 stream connections with all the TiDB
+servers.
+
+![TSO Follower Proxy](https://i.imgur.com/WoB9YN9.png)
```

Comment:
BTW, should we move the image to the media directory?

Reply:
Updated. PTAL.